src/codec.rs

```diff
@@ -781,10 +785,32 @@ impl Decode for Vec<u8> {
 	fn decode<I: Input>(input: &mut I) -> Option<Self> {
 		<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 			let len = len as usize;
-			let mut vec = vec![0; len];
-			if input.read(&mut vec[..len]) != len {
-				None
+
+			if len < MAX_PREALLOCATION_SIZE {
+				let mut vec = vec![0; len]; // len is ok here
+				if input.read(&mut vec[..]) != len {
```

Comment:
I believe that's wrong. `std::io::Read::read` can just read less bytes in case they are not available yet (like when you read directly from network or a slow disk) - we should try in a loop until we get `0` like we do with dynamic allocation. I'd unify the two codes and just have a bool flag to control if we are allowed to go over current capacity.

Reply:
This is not `std::io::Read`.

Reply:
Or do you mean because we implement `std::io::Read` for `Input`. We should probably drop this implementation...
Even `0` does not mean that we read until the end of a file in certain situations.. And I clearly don't want to loop here.

Reply:
I created an issue for this. Ty.

Reply:
Okay, looked again into the implementation :sweat_smile: 
We use https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact for the `std::io::Read` impl. So, we are safe and don't need to do that.

Reply:
Yeah, that was my point. Thanks!